domain/removal/service/application.go

```diff
@@ -0,0 +1,140 @@
+
+package service
+
+import (
+	"context"
+	"time"
+
+	coreapplication "github.com/juju/juju/core/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/domain/life"
+	"github.com/juju/juju/domain/removal"
+	removalerrors "github.com/juju/juju/domain/removal/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// ApplicationState describes retrieval and persistence
+// methods specific to application removal.
+type ApplicationState interface {
+	// ApplicationExists returns true if a application exists with the input
+	// name.
+	ApplicationExists(ctx context.Context, appUUID string) (bool, error)
+
+	// EnsureApplicationNotAlive ensures that there is no application identified
+	// by the input name, that is still alive.
+	EnsureApplicationNotAlive(ctx context.Context, appUUID string) (err error)
+
+	// ApplicationScheduleRemoval schedules a removal job for the application
+	// with the input name, qualified with the input force boolean.
+	ApplicationScheduleRemoval(ctx context.Context, removalUUID, appUUID string, force bool, when time.Time) error
+
+	// GetApplicationLife returns the life of the application with the input
+	// UUID.
+	GetApplicationLife(ctx context.Context, appUUID string) (life.Life, error)
+
+	// DeleteApplication removes a application from the database completely.
+	DeleteApplication(ctx context.Context, appUUID string) error
+}
+
+// RemoveApplication checks if a application with the input name exists.
+// If it does, the application is guaranteed after this call to be:
+// - No longer alive.
+// - Removed or scheduled to be removed with the input force qualification.
+// The input wait duration is the time that we will give for the normal
+// life-cycle advancement and removal to finish before forcefully removing the
+// application. This duration is ignored if the force argument is false.
+// The UUID for the scheduled removal job is returned.
+func (s *Service) RemoveApplication(
+	ctx context.Context,
+	appUUID coreapplication.ID,
+	force bool,
+	wait time.Duration,
+) (removal.UUID, error) {
+	exists, err := s.st.ApplicationExists(ctx, appUUID.String())
+	if err != nil {
+		return "", errors.Errorf("checking if application %q exists: %w", appUUID, err)
+	} else if !exists {
+		return "", errors.Errorf("application %q does not exist", appUUID).Add(applicationerrors.ApplicationNotFound)
+	}
+
+	// Ensure the application is not alive.
+	if err := s.st.EnsureApplicationNotAlive(ctx, appUUID.String()); err != nil {
+		return "", errors.Errorf("application %q: %w", appUUID, err)
+	}
+
+	if force {
+		if wait > 0 {
+			// If we have been supplied with the force flag *and* a wait time,
+			// schedule a normal removal job immediately. This will cause the
+			// earliest removal of the application if the normal destruction
+			// workflows complete within the the wait duration.
+			if _, err := s.applicationScheduleRemoval(ctx, appUUID, false, 0); err != nil {
```

Comment:
Shouldn't wait be passed to the scheduled removal?
```suggestion
			if _, err := s.applicationScheduleRemoval(ctx, appUUID, false, wait); err != nil {
```

Reply:
What @jack-w-shaw said